### DISCUSSION TITLE
Add support to Next/Image to specify a subset of device and image sizes to reduce srcset bloat

### DISCUSSION THREAD

**Discussion by chriseagle** [vercel/next.js#41731] [Ideas]
### Describe the feature you'd like to request

The current implementation of configuring `imageSizes` and `deviceSizes` in next.config.js allows us to define sizes which Next will generate images for. This currently adds every size specificied to every `img` tag that Next/Image ultimately renders.

**next.config.js example**
```
const nextConfig = {
	images: {
		imageSizes: [96, 154], 
		deviceSizes: [320, 360, 390, 414, 768, 960, 1024, 1440, 1920],
	}
};
```

The final HTML output results in a bloated `srcset` attribute on every img tag in the page. As a source set URL will be generated and provided for each size defined.

When building for performance and optimising images, we may want to define image sizes so that Lighthouse does not penalise us for [incorrectly sized images](https://web.dev/uses-responsive-images/).

Take this example as a good use case, when defining this sizes for a large hero image. 

We may want to define a 960, 1440, and 2880, to cover high resolution(3x) mobile (960), desktop (1440) and high resolution desktop (2880), if the project is focused on high quality media, and is found in a media rich page, this hero element will only occur once, typically above the fold. However with the current API, every image tag on the page will now include srcset entries for 960, 1440 and 2880.

When analysing pages on a large media rich publishing project, we noted that our image source set values account for over a **quarter** of our total page weight.

![Screenshot 2022-10-21 at 13 03 49](https://user-images.githubusercontent.com/4096392/197191528-e911ebd6-c840-447d-8b4f-4c2d917fa5f2.png)

I note that next/future/image details that `sizes` attribute can now be used generate the image source set, however I have reservations that this currently as optimised as it can be.  Providing an absolute value to Next/Future/Image, results in even greater source set bloat at present.  Please see alternatives considered below.

When delivering modern image formats as such `webp` and `avif`, and these formats have been optimised, a smaller source set is highly desirable to achieve even greater page performance gains.

### Describe the solution you'd like

The solution I would propose would be for `Next/Image` to accept a new prop named `allowedImageSizes` or similar (naming things is hard).

This prop would accept an array of images sizes, which would need to match with the defined and final concatenated list from next.config.js to be considered valid. 

`<Image
      src="/example.png"
      allowedImageSizes={[960, 1440, 2880]}
      sizes="100vw"
    />`

This allow us to target single Next/Image components to only populate the image source set with the supplied sizes and with a width descriptor for each source set entry, matching that of the `sizes` attribute passed in. 

The final output `img` tag could look like so

`<img sizes="100vw" srcset="/_next/image?url=/image.jpg&w=960&q=75 960w, "/_next/image?url=/image.jpg&w=1440&q=75 1440w, "/_next/image?url=/image.jpg&w=2880&q=75 2880w" src="/_next/image?url=/image.jpg&w=2880&q=75 2880w" width="2400" height="1443" decoding="async" data-nimg="future" loading="lazy" style="color:transparent">`

This feature would allow further flexibility for developers to specify a limited source set for individual image tags, helping to improve overall page performance and reducing current source set bloat. 

### Describe alternatives you've considered

We explored abandoning Next/Image and using a third party image provider, we also explored the new Next/Future/image and its support for generating the source set from the sizes provided, however we feel this solution falls short of desired result and is not ready for production.

Using `Next/Future/Image` with the hero example above and providing it a simple `sizes="100vw"` value, the final `img` tag is still populated with a wide range source set values which are ultimately not required. 

`<img sizes="100vw" srcset="/_next/image?url=/image.jpg&w=640&q=75 640w, /_next/image?url=/image.jpg&w=750&q=75 750w, /_next/image?url=/image.jpg&w=828&q=75 828w, /_next/image?url=/image.jpg&w=1080&q=75 1080w, /_next/image?url=/image.jpg&w=1200&q=75 1200w, /_next/image?url=/image.jpg&w=1920&q=75 1920w, /_next/image?url=/image.jpg&w=2048&q=75 2048w, /_next/image?url=/image.jpg&w=3840&q=75 3840w" src="/_next/image?url=/image.jpg&w=3840&q=75" width="2400" height="1443" decoding="async" data-nimg="future" loading="lazy" style="color:transparent">`

The source set bloat is even worse when providing an absolute width such as `sizes="1000px"`

`<img sizes="1000px" srcset="/_next/image?url=/image.jpg&w=16&q=75 16w, /_next/image?url=/image.jpg&w=32&q=75 32w, /_next/image?url=/image.jpg&w=48&q=75 48w, /_next/image?url=/image.jpg&w=64&q=75 64w, /_next/image?url=/image.jpg&w=96&q=75 96w, /_next/image?url=/image.jpg&w=128&q=75 128w, /_next/image?url=/image.jpg&w=256&q=75 256w, /_next/image?url=/image.jpg&w=384&q=75 384w, /_next/image?url=/image.jpg&w=640&q=75 640w, /_next/image?url=/image.jpg&w=750&q=75 750w, /_next/image?url=/image.jpg&w=828&q=75 828w, /_next/image?url=/image.jpg&w=1080&q=75 1080w, /_next/image?url=/image.jpg&w=1200&q=75 1200w, /_next/image?url=/image.jpg&w=1920&q=75 1920w, /_next/image?url=/image.jpg&w=2048&q=75 2048w, /_next/image?url=/image.jpg&w=3840&q=75 3840w" src="/_next/image?url=/image.jpg&w=3840&q=75" width="2400" height="1443" decoding="async" data-nimg="future"  loading="lazy" style="color:transparent">`

I can't fathom a reason why `next/future/image` now feels a version at 16px wide is required, and the source set now includes a complete entry list of all default values from `next.config.js` of both `deviceSizes` and `imageSizes`. Which hopefully goes someway towards the reason why I'm suggesting this feature.

Thanks !

**Comment by karlhorky** [comment]
@balazsorban44 are you sure that this should be changed into a discussion? This last bit about using a fixed `1000px` and getting **more** HTML generated seems like a bug to me... 🤔

**Reply by karlhorky** [reply]
Maybe this is reported over here already: https://github.com/vercel/next.js/issues/27547

**Reply by chriseagle** [reply]
@karlhorky Would agree on the bug front, thanks for the link.

**Reply by karlhorky** [reply]
Seems like others are having similar issues with `srcSet` and `next/image`, although not 100% the same as the report above: https://github.com/vercel/next.js/issues/37265

**Comment by cramforce** [comment]
If the `sizes` attribute is correct, then this will never render too large images (there is a Chrome bug where it picks large images for very high pixel densities but that should be fixed soon and it is minor and Lighthouse doesn’t bemoan it). If you have a sample page that triggers the Lighthouse test I'm happy to take a look. If it is a false-positive we can get it fixed in Lighthouse.

Could it be that the `sizes` attribute wasn't correct when this was originally tested? If I see `sizes=100vw` then that often indicates that this is likely not correct on desktop.

That leaves us with `srcset` bloat which is a bit annoying but gzips/brotlis away entirely so it doesn’t hurt on the wire. Something like `allowedImageSizes` would help with that but it would likely be very rarely used and make the docs more complicated so I'm hesitant to add such a feature. Maybe using a custom loader that replicates the default output but filters the sizes would help in your case.

**Reply by chriseagle** [reply]
Thanks for the reply. I agree on the HTTP compression front, however, it still feels like the wrong approach, like throwing a full-page cache up in front of a slow site to plaster over the cracks.

From what I can see reading the docs, custom loaders don't appear to break you out of the sizes/srcset defaults that Next handles. If you have documentation around this, I would like to read it.

On the `sizes` attribute with the new next/image, perhaps absolute value support hasn't been added? As the docs now state:

_If the sizes property includes sizes such as 50vw, which represent a percentage of the viewport width, then the source set is trimmed to not include any values which are too small to ever be necessary._

However, the result below is what is generated when passing in `1000px`, a complete `srcset` for every default `deviceSizes` and `imageSizes` down to 16(?!). This can be reduced by defining our own sizes in the `next.config.js`, however this then restricts all images which do require a larger source set.

![Screenshot 2022-11-07 at 08 57 39](https://user-images.githubusercontent.com/4096392/200532258-40ae6e5f-b283-4a53-b85a-683a24033438.png)

There is no way to break out of the defaults, which is why I requested the possibility of this feature for more granular control.
